src/cargo/ops/cargo_clean.rs

```diff
@@ -57,43 +59,47 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
     if opts.spec.is_empty() {
         return rm_rf(&target_dir.into_path_unlocked(), config);
     }
-    let (packages, resolve) = ops::resolve_ws(ws)?;
-
-    let interner = UnitInterner::new();
     let mut build_config = BuildConfig::new(config, Some(1), &opts.target, CompileMode::Build)?;
     build_config.requested_profile = opts.requested_profile;
     let target_data = RustcTargetData::new(ws, build_config.requested_kind)?;
-    let bcx = BuildContext::new(
-        ws,
-        &packages,
-        opts.config,
-        &build_config,
-        profiles,
-        &interner,
-        HashMap::new(),
-        target_data,
-    )?;
-    let requested_features = RequestedFeatures::new_all(true);
+    let resolve_opts = ResolveOpts::everything();
     let specs = opts
         .spec
         .iter()
         .map(|spec| PackageIdSpec::parse(spec))
         .collect::<CargoResult<Vec<_>>>()?;
-    let features = FeatureResolver::resolve(
+    let ws_resolve = ops::resolve_ws_with_opts(
```

Comment:
How come this changed from `resolve_ws` to `resolve_ws_with_opts`?

I feel like we keep trying to move everything over to `resolve_ws`, but things seem to oscillate between the two methods over tiem?

Reply:
I switched it so that it could reuse the code for building the feature resolver and downloading dependencies.  Otherwise, that would need to be duplicated here if it continued to use `resolve_ws`.

FWIW, I'd really like to completely change how `cargo clean -p` works some day.  This technique of trying to guess what the filename hash is just too cumbersome and fragile, and doesn't always work.  I hope that we can get some on-disk metadata to sufficiently track it so that `cargo clean` doesn't need to do all this silliness.

Reply:
Hm ok. We may want to add a helper method to "download everything necessary" at some point, but seems reasonable enough for now.